### ISSUE TITLE
incorrect WARNING - invalid cast - the types do not have a common subtype in NTI

### ISSUE THREAD

**Issue by ghosttie** [google/closure-compiler#2092]
I'm getting this compile error:

```
filevision/ui/field/ComboboxField.js:450: ERROR - invalid cast - the types do not have a common subtype
from: Array<string>|Date|filevision.net.API.DisplayObject|boolean|number|string
to  : Array<filevision.net.API.DisplayObject>
        var /** !Array<!filevision.net.API.DisplayObject> */ objects = /** @type {!Array<!filevision.net.API.DisplayObject>} */ (response.result);
                                                                                                                                ^
```

The type annotation for result is:

```
 * @type {string|!Date|number|boolean|!filevision.net.API.DisplayObject|!Array<!filevision.net.API.DisplayObject>|!Array<!filevision.net.API.TableRow>|!Array<string>}
```

so casting it as a !Array<!filevision.net.API.DisplayObject> should be fine.

I can't create an exact reproducer but [this](https://closure-compiler-debugger.appspot.com/#input0%3Dvar%2520%252F**%2520string%257C!Date%257Cnumber%257Cboolean%257C!filevision.net.API.DisplayObject%257C!Array%253C!filevision.net.API.DisplayObject%253E%257C!Array%253C!filevision.net.API.TableRow%253E%257C!Array%253Cstring%253E%2520*%252F%2520result%2520%253D%2520%255B%255D%253B%250A%250Avar%2520objects%2520%253D%2520%252F**%2520%2540type%2520%257B!Array%253C!filevision.net.API.DisplayObject%253E%257D%2520*%252F%2520%28result%29%253B%26input1%26conformanceConfig%26externs%3D%252F**%2520%2540const%2520*%252F%250Avar%2520filevision%2520%253D%2520%257B%257D%253B%250A%252F**%2520%2540const%2520*%252F%250Afilevision.net%2520%253D%2520%257B%257D%253B%250A%252F**%2520%2540const%2520*%252F%250Afilevision.net.API%2520%253D%2520%257B%257D%253B%250A%250A%252F**%2520%2540constructor%2520*%252F%250Afilevision.net.API.DisplayObject%253B%250A%250A%252F**%2520%2540constructor%2520*%252F%250Afilevision.net.API.TableRow%253B%26refasterjs-template%26includeDefaultExterns%3D1%26CHECK_SYMBOLS%3D1%26CHECK_TYPES%3D1%26CHECK_TYPES_NEW_INFERENCE%3D1%26CLOSURE_PASS%3D1%26LANG_IN_IS_ES6%3D1%26MISSING_PROPERTIES%3D1%26PRESERVE_TYPE_ANNOTATIONS%3D1%26PRETTY_PRINT%3D1%26TRANSPILE%3D1) seems pretty close.

**Comment by dimvar** [comment]
Small repro:

https://closure-compiler-debugger.appspot.com/#input0%3D%252F**%2520%2540constructor%2520*%252F%250Afunction%2520Foo()%2520%257B%257D%250A%250A%252F**%2520%2540constructor%2520*%252F%250Afunction%2520Bar()%2520%257B%257D%250A%250A%252F**%2520%2540type%2520%257B!Array%253C!Foo%253E%257C!Array%253C!Bar%253E%257C!Array%253Cstring%253E%257D%2520*%252F%250Avar%2520result%2520%253D%2520%255B%255D%253B%250A%250Avar%2520objects%2520%253D%2520%252F**%2520%2540type%2520%257B!Array%253C!Foo%253E%257D%2520*%252F%2520(result)%253B%26input1%26conformanceConfig%26externs%26refasterjs-template%26includeDefaultExterns%3D1%26CHECK_SYMBOLS%3D1%26CHECK_TYPES%3D1%26CHECK_TYPES_NEW_INFERENCE%3D1%26CLOSURE_PASS%3D1%26LANG_IN_IS_ES6%3D1%26MISSING_PROPERTIES%3D1%26PRESERVE_TYPE_ANNOTATIONS%3D1%26PRETTY_PRINT%3D1%26TRANSPILE%3D1

**Comment by vobruba-martin** [comment]
A walkaround for this:

```javascript
var /** string|!Date|number|boolean|!filevision.net.API.DisplayObject|!Array<(!filevision.net.API.DisplayObject|!filevision.net.API.TableRow|string)> */ result = [];
```
